database/src/ops/block.rs

```diff
@@ -1,89 +1,327 @@
-    todo!()
-}
+    // Block Info.
+    tables.block_infos_mut().put(
+        &height,
+        &BlockInfo {
+            timestamp: block.header.timestamp,
+            total_generated_coins: generated_coins,
+            cumulative_difficulty,
+            block_hash,
+            cumulative_rct_outs, // TODO: should this be a dummy value?
+            // INVARIANT: #[cfg] @ lib.rs asserts `usize == u64`
+            weight: weight as u64,
+            long_term_weight: long_term_weight as u64,
+        },
+    )?;
 
-/// TODO
-pub fn get_block_weight() {
-    todo!()
-}
+    // Block blobs.
+    tables
+        .block_blobs_mut()
+        .put(&height, &StorableVec(block_blob))?;
 
-/// TODO
-pub fn get_block_already_generated_coins() {
-    todo!()
-}
+    // Block heights.
+    tables.block_heights_mut().put(&block_hash, &height)?;
 
-/// TODO
-pub fn get_block_long_term_weight() {
-    todo!()
-}
+    // Transaction & Outputs.
+    for tx in txs {
+        let tx: &Transaction = &tx.tx;
+        add_tx(tx, tables)?;
 
-/// TODO
-pub fn get_block_timestamp() {
-    todo!()
-}
+        // ^
+        // Everything above is adding tx data.
+        // Everything below is for adding input/output data.
+        // v
 
-/// TODO
-pub fn get_block_cumulative_rct_outputs() {
-    todo!()
+        // Is this a miner transaction?
+        // Which table we add the output data to depends on this.
+        // <https://github.com/monero-project/monero/blob/eac1b86bb2818ac552457380c9dd421fb8935e5b/src/blockchain_db/blockchain_db.cpp#L212-L216>
+        let mut miner_tx = false;
+
+        for inputs in &tx.prefix.inputs {
+            match inputs {
+                // Key images.
+                Input::ToKey { key_image, .. } => {
+                    add_key_image(tables.key_images_mut(), key_image.compress().as_bytes())?;
+                }
+                // This is a miner transaction, set it for later use.
+                Input::Gen(_) => miner_tx = true,
+            }
+        }
+
+        // Output bit flags.
+        // Set to a non-zero bit value if the unlock time is non-zero.
+        let output_flags = match tx.prefix.timelock {
+            Timelock::None => 0x0000_0000,
+            Timelock::Block(_) | Timelock::Time(_) => 0x0000_0001,
+        };
+
+        // Output data.
+        for (i, output) in tx.prefix.outputs.iter().enumerate() {
+            let tx_idx = get_num_tx(tables.tx_ids_mut())?;
+            let key = *output.key.as_bytes();
+
+            // RingCT (v2 transaction) miner outputs.
+            if miner_tx && tx.prefix.version == 2 {
+                add_rct_output(
+                    &RctOutput {
+                        key,
+                        height,
+                        output_flags,
+                        tx_idx,
+                        commitment: todo!(), // Zero dummy commit?
+                    },
+                    tables.rct_outputs_mut(),
+                )?;
```

Comment:
What is `commitment` in the v2 miner transaction case?

Reply:
```cpp
/* (blockchain_db/blockchain_db.cpp, l.232) */
    // miner v2 txes have their coinbase output in one single out to save space,
    // and we store them as rct outputs with an identity mask
    if (miner_tx && tx.version == 2)
    {
      cryptonote::tx_out vout = tx.vout[i];
      rct::key commitment = rct::zeroCommit(vout.amount);
      vout.amount = 0;
      amount_output_indices[i] = add_output(tx_hash, vout, i, tx.unlock_time,
        &commitment);
    }
```
Zero commitment indeed.

Reply:
it's:

```math
Commitment = G +  H*ammount
```

Where G is the base point and H is from monero-serai, amount is the output amount.

Monero implements a look up table for common these values: https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/ringct/rctOps.cpp#L322

For now we can just compute them all on the fly